### ISSUE TITLE
GIFs animate slowly with Glide 4.2

### ISSUE THREAD

**Issue by moxie0** [bumptech/glide#2471]
Here is a screen capture of two gifs playing w/ Glide 3.7:
![slowgif](https://user-images.githubusercontent.com/512439/31471903-2d611a6a-aea0-11e7-9f92-5ab4d83dc3cb.gif)

Here is a screen capture of the same two gifs playing w/ Glide 4.2:
![fastgif](https://user-images.githubusercontent.com/512439/31472012-f0ca007a-aea0-11e7-882e-2b4ef259a954.gif)

`````
      GlideApp.with(context)
              .load(url)
              .diskCacheStrategy(DiskCacheStrategy.ALL)
              .into(view);
`````

This happens with any gif that I've tried, here's an example:
https://media.giphy.com/media/3ohjV3cQ9lvPeCVLOg/giphy.gif

**Glide Version**: 4.2.0
**Integration libraries**: None
**Device/Android Version**: Nexus 5, OnePlus One, Pixel

**Comment by lbx2015** [comment]
same as me. Can you give me some advise about the hand on code to load gif with 4.2.0?
RequestOptions o = new RequestOptions();
       Glide.with(h.imageView3.getContext())
.load("https://timgsa.baidu.com/timg?image&quality=80&size=b9999_10000&sec=1508322038285&di=e0c474540228565f1271de3b4b2c56e7&imgtype=0&src=http%3A%2F%2Fwww.lia-edu.com%2Fupload%2Fimage%2F20170717%2F20170717151619_8233.gif").apply(o.diskCacheStrategy(DiskCacheStrategy.RESOURCE)).into(h.imageView3);

**Comment by sjudd** [comment]
I'm not able to reproduce this on a Nexus 5X on 7.1.1 if I download the gifs you've pointed to. The GIFs are however pretty small.

Can you attach the files you're having trouble with directly to the issue to make sure we're testing using the same exact data?

Can you also provide the XML or at least the dimensions of the view you're loading into?

**Comment by moxie0** [comment]
Hmm, I think I must have made a mistake in my original issue report. I just tried building a sample app with a jarjar'd version of 3.7 and an unmodified 4.2 together in the same app which loads the same GIF into two adjacent ImageViews using two different versions of Glide, and I'm also not able to reproduce this.

What I am seeing, however, is that GIFs render in slow-motion with both Glide 3.7 and 4.2 on Android 5.0 and Android 6.0 devices (a Nexus 5 and OnePlus One, respectively), but render at normal speed on an Android 8.0 device (Pixel).

I tried a sample app with a single 3.7 or 4.2 production dependency, loading a single GIF, and the results are the same. Things render fine (with both 3.7 and 4.2) on an Android 8.0 device, but not an Android 5.0 or 6.0 device.

Here's a sample layout I'm using:

`````
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout
        xmlns:android="http://schemas.android.com/apk/res/android"
        android:orientation="vertical"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        android:layout_gravity="center">
        
        <ImageView
                android:id="@+id/image"
                android:layout_gravity="center"
                android:layout_width="210dp"
                android:layout_height="210dp"/>

</LinearLayout>
`````

The Glide 4.2 code:

`````
    Glide.with(this)
         .load(R.raw.giphy)
         .apply(RequestOptions.diskCacheStrategyOf(DiskCacheStrategy.NONE))
         .into((ImageView)findViewById(R.id.image));
`````

An attached GIF I've tested with:
![giphy](https://user-images.githubusercontent.com/512439/32188373-da40378e-bd64-11e7-88f7-b6c29b81760d.gif)

**Comment by sjudd** [comment]
Thanks for the follow up. 

That makes sense. The [GIF decoder][1] we use isn't necessarily as fast or optimized as it could be. We've fixed a few cases where it was slower than it needed to be, but there may be more relatively simple changes we could make. Otherwise it might require a more fundamental restructuring.

One other idea we've been looking in to is adding support for rastermill/framesequence or another third party gif decoder as an integration library.

[1]: https://github.com/bumptech/glide/blob/477e6608c16a8c39d91948c9d410ab04efc13e13/third_party/gif_decoder/src/main/java/com/bumptech/glide/gifdecoder/StandardGifDecoder.java#L417

**Comment by ssessa** [comment]
I ran into this issue on a Nexus 5 running 6.0.1. 

When running the debug Build Variant gifs perform poorly. 

Once I switched to release, gifs began performing as expected. Hope this helps out...

**Comment by sjudd** [comment]
I haven't found anything huge so far but two things worth noting:

1. In Glide 4.3.0, the number of threads available for GIFs decreased to 2 or 1. These threads are used only for GIF frames which avoids delays loading frames for network operations, but reduced the number of threads available for GIF frames, especially on high end devices. Limiting the number of concurrently playing GIFs will help. I'll also add an API in Glide 4.4.0 to allow you to modify the number of threads used for animations.
2. In Glide 4.4.0, I'll add an (experimental) API called ``clearOnDetach`` (name may change, suggestions welcome). ``clearOnDetach`` will free up resources used by ``ViewTarget``s when the ``View`` is detached from it's window and automatically restart the request if the ``View`` is ever reattached. That method is primarily meant for reducing memory usage, but it will also help reduce the number of running GIFs by pausing those as soon as they go off screen. I've seen this help significantly in RecyclerView, which doesn't set visibility for off screen views and doesn't call its ``RecyclerListener`` for every view that's off screen. I may also try to just build this in to ``ViewTarget`` so that it happens automatically.

More to come as I find it. If anyone else has any suggestions or can reproduce any particularly bad cases, I'd welcome the help.

**Comment by sjudd** [comment]
@ssessa Are you using proguard? Do you have optimizations enabled? If you disable proguard but otherwise leave you release variant the same, do you still see GIF performance improve on those builds?

**Comment by sjudd** [comment]
I've pushed a series of (hopefully) safe changes. Referencing the intermediate arrays in local variables seems to make a really large difference on API 23 devices, though I don't see the same difference on API 26. The rest of the changes are much smaller improvements. 

I have a couple of larger fixes including automatic pausing of detached animatables and an option to cache color table indices for all GIF frames (one byte per changed pixel per frame vs four bytes per pixel for all pixels per frame if we cache the entire frame). I may hold off on those until 4.5 to give us some more time to test. 

GIFs are difficult to test and there's lots of expected but not to spec behavior that can make things that seem like they ought to be safe unsafe.

**Comment by delacrixmorgan** [comment]
Realised that we are looking for the wrong place for the GIF issue. Problem isn't with Glide, but it how Android is recognising the GIF file. Try this.

> Add this url into your Glide load, and it will load nicely. Although, in reality it's 7.2MB.
> https://media.giphy.com/media/3o6Zt7YExBfE89p6Yo/giphy.gif

In fact, because the API that Giphy uses for that image was specify the `content-type` as `image/gif`. Thus, somehow made Android recognise GIF file and handles them respectively.

**TLDR**
Specifying `content-type` as `image/gif` will make it work. Simple indicator enter your link in your browser and see whether it hosts the image on the page or prompt you to download.

- If the GIF is host on the page, it'll work fine.
- If the GIF prompts for a download, it'll have that lag we all have been dreading about.

**Comment by KimiChiu** [comment]
Hi, 

I have the same problem too.
In Glide 3, we fix this problem by this way.

    builder.setDiskCacheService(new FifoPriorityThreadPoolExecutor(4));
    builder.setResizeService(new FifoPriorityThreadPoolExecutor(4));

But we can't found these APIs in Glide 4.

**Comment by sjudd** [comment]
@KimiChiu http://bumptech.github.io/glide/javadocs/460/com/bumptech/glide/GlideBuilder.html#setAnimationExecutor-com.bumptech.glide.load.engine.executor.GlideExecutor-

**Comment by Eggplant-Cui** [comment]
For me
I did some invesgate
Obviously this is related to hardware
I tested a high resolution gif (1920*1080 120 frames 867KB) on several devices
Perfect performance on:
 XiaoMi 6 Qualcomm 845 published 2017
OnePlus 5 Qualcomm 845 published 2017
Normal performance on:
Samsung S5 Qualcomm 800 published 2014
Bad performance on:
Oppo R7 plus Qualcomm 615 published 2015

The test is not only on glide , i also try to play this gif with the default way the device offered.

But if i choose another gif (400*300, 7.2MB), it performance perfectly on all devices above.

So try to compress your gif or choose play it as movie will partly solve this problem.

Sorry for my poor English, I'll repeat my words in Chinese

gif动画帧率低的问题应该不只是glide的问题，各位如果能把设备信息和gif信息一起贴上来可能会更快的明确问题是关于设备还是这个库。
我的任务是为app添加开屏动画，动画尺寸是1080p，120帧，帧率60，文件大小867kb
在小米6，一加5上面完美播放
三星s5略有延迟，大概2.1-2.2秒才能播放完毕
oppo r7 plus 需要4-5秒才能播放完毕
而且调用系统相册播放gif也是相同的结果
而我换了一个400*300分辨率，7.2mb大小的gif文件，无论是用glide还是用系统的相册播放，都很流畅。所以我猜测这个问题应该是cpu解码gif的速度跟不上播放速度导致的

最终我的解决方案是用movie来播放gif，这样可能会导致丢帧，但是至少时间长度是一致的

**Comment by perracolabs** [comment]
Any plans to add a third-party decoder to improve gif playback?

**Comment by andrefrsousa** [comment]
@sjudd Any update on this?

**Comment by lucasgriotto** [comment]
@andrefrsousa can you upload the code please?
Thanks

**Comment by andrefrsousa** [comment]
@lucasgriotto very simple:
**Glide**
```
Glide.with(holder.binding.imageView)
                .asGif()
                .load(image[position])
                .into(holder.binding.imageView)
                .clearOnDetach()
```

**Fresco**

```
        val hierarchy = holder.binding.imageView.hierarchy
        hierarchy.fadeDuration = 0

        holder.binding.imageView.hierarchy = hierarchy
        holder.binding.imageView.controller = Fresco.newDraweeControllerBuilder()
                .setImageRequest(ImageRequest.fromUri(image[position]))
                .setAutoPlayAnimations(true)
                .build()

```

**Comment by lucasgriotto** [comment]
@andrefrsousa thanks. I don't know what is wrong I can't get it working properly.

I am using Glide.

```
 implementation 'com.github.bumptech.glide:glide:4.7.1'
 annotationProcessor 'com.github.bumptech.glide:compiler:4.7.1'

<ImageView
        android:id="@+id/fra_home_img_loading_burger"
        android:layout_width="match_parent"
        android:layout_height="match_parent" />

mImgLoading = view.findViewById(R.id.fra_home_img_loading_burger);


 Glide.with(mImgLoading).asGif().load("https://media.giphy.com/media/3o6Zt7YExBfE89p6Yo/giphy.gif")
.into(mImgLoading).clearOnDetach();

```

**Comment by perracolabs** [comment]
The reason Fresco is faster is because it caches all the GIF frames, and in addition uses some native help for decoding. Source code here:

[Fresco's GIF decoder](https://github.com/facebook/fresco/blob/2116559bad1dcd03b4a86be1c9863eb5685f6b8b/animated-gif-lite/src/main/java/com/facebook/animated/giflite/GifDecoder.java)

Notice how all the frames get cached.  And for a fast decoding it uses the [android.graphics.Movie](https://developer.android.com/reference/android/graphics/Movie) class.
If you check the source of [**Movie.decodeStream()**](https://github.com/aosp-mirror/platform_frameworks_base/blob/e80b45506501815061b079dcb10bf87443bd385d/graphics/java/android/graphics/Movie.java), you’ll see how this one uses a native method for decoding: **nativeDecodeStream**

Glide doesn’t cache GIF frames to reduce memory usage, plus the decoding is done purely in Java.

I think an improvement would be to give an optional setting to cache GIF frames in memory.  It makes sense to not cache them by default if the GIFs are seen in a RecyclerView with lots of other images. But once a user selects/opens one specific GIF, it would be better to use an optimized version with memory cached frames. Maybe as a caution, it could also have a parameter to specify the size for cached memory frames.

A second optimization would be to also use the [Movie](https://developer.android.com/reference/android/graphics/Movie) class, or the [AnimatedImageDrawable](https://developer.android.com/reference/android/graphics/drawable/AnimatedImageDrawable) starting at API 28, as Movie class has just been deprecated.

**Comment by githubzjh** [comment]
@sjudd  Do you have some plans to fix the issue?

**Comment by Zhuinden** [comment]
Maybe this spares some time for people in the future, so what I did was ditch Glide in favor of Fresco because it loads GIFs fast.

So what I did is add Fresco:

```
    implementation 'com.facebook.fresco:fresco:1.9.0'
    // For animated GIF support
    implementation 'com.facebook.fresco:animated-gif:1.9.0'
    // For WebP support, including animated WebP
    implementation 'com.facebook.fresco:animated-webp:1.9.0'
    implementation 'com.facebook.fresco:webpsupport:1.9.0'
```

And the `Fresco.initialize(context)` call and stuff that it needs, and also

``` kotlin
class GifDraweeView : SimpleDraweeView {
    private var drawableResource = 0

    constructor(context: Context) : super(context) {
        init(context, null, 0)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context, attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context, attrs, defStyleAttr)
    }

    private fun init(context: Context, attributeSet: AttributeSet?, defStyle: Int) {
        if (attributeSet != null) {
            val a: TypedArray? = when {
                defStyle != 0 -> context.obtainStyledAttributes(attributeSet, R.styleable.GifDraweeView, defStyle, 0)
                else -> context.obtainStyledAttributes(attributeSet, R.styleable.GifDraweeView)
            }
            drawableResource = a!!.getResourceId(R.styleable.GifDraweeView_gifResourceDrawable, 0)
            a.recycle()


            if (!isInEditMode) {
                if (drawableResource != 0) {
                    val controller = Fresco.newDraweeControllerBuilder()
                        .setUri(UriUtil.getUriForResourceId(drawableResource))
                        .setAutoPlayAnimations(true)
                        .build()
                    setController(controller)
                }
            }
        }
    }
}
```

and

```
<declare-styleable name="GifDraweeView">
        <attr name="gifResourceDrawable" format="reference"/>
</declare-styleable>
```

And Fresco GIF rendering is not slow, so it's ok

**Comment by AndroidDeveloperLB** [comment]
Any news about this?

@Zhuinden Can you please share a full sample of just that? I also wonder how to go over each frame instead of just playing them directly into a View.

**Comment by Zhuinden** [comment]
I'm pretty sure you have to implement yourself what `GifFrameLoader` is doing in order to go through frames one by one instead of playing them directly into a View.

Check the GifDrawable source.

I didn't try using Glide 3.8.0 (or 3.7.0) to see if performance is better, but that might be an option too...

**Comment by AndroidDeveloperLB** [comment]
@Zhuinden I mean "can you please share a full sample of using Fresco?" And in the same context "I also wonder how to go over each frame instead of just playing them directly into a View, in Fresco"

**Comment by Zhuinden** [comment]
oh I think in Fresco you'd have even tougher luck going frame by frame, I have no idea.

**Comment by delacrixmorgan** [comment]
GIPHY wrote a blog post explaining the details of how Fresco handles GIF better and to quote the post.

> Developers made it clear that performance can’t be improved, unless they rewrite their rendering engine from scratch.

https://engineering.giphy.com/giphy-android-app-and-fresco/

**Comment by idish** [comment]
Pretty major issue, this should be fixed somehow in the next release!

**Comment by skyfit2** [comment]
I have the same problem , i recomended  change to Fresco , its more fast and complete compared to Glide

**Comment by takaoandrew** [comment]
@skyfit2 Do you know if Fresco allow playing GIFs from a url like Glide does?

**Comment by AndroidDeveloperLB** [comment]
@takaoandrew  For the size of this library, it better be....

**Comment by nirukk52** [comment]
Adding crossfade on local image drawable worked for me.

```
Glide.with(context).load(stringImage).asGif()
   .crossFade().diskCacheStrategy(DiskCacheStrategy.SOURCE).into(postImage);
```

**Comment by andrewcking** [comment]
Also experiencing this on most devices I have tested in on from Galaxy s10 to nexus 5.  It is much more subtle on higher end devices but is still noticeable.

**Comment by Zhuinden** [comment]
Have you tried with Glide 3.7.2?

**Comment by andrewcking** [comment]
Wow..... just tried doing some rewrites to go back to 3.7.2 and gif performance dramatically improved. This makes me chuckle because they cited improved Gif performance as one of the reasons to upgrade to 4. I have been sticking with Glide 4.2 because performance got a lot worse in the 4.3 update though I have been testing each version afterwards to see if it gets better and it hasn't.

Here is what I have found:
Glide 4.3-4.9: Terrible gif performance on all devices
Glide 4.2: Acceptable gif performance on most devices
Glide 3.7.2: Great gif performance (so far in my tests)

**Comment by AndroidDeveloperLB** [comment]
@andrewcking Can you please show a sample, so that I could see the difference?

**Comment by andrewcking** [comment]
Visually? There are some examples above but here is what I am seeing in terms of performance difference between 3.7 and 4.9 [download.zip](https://github.com/bumptech/glide/files/3146008/glide.gif.3-7.vs.4-9.zip)

**Comment by AndroidDeveloperLB** [comment]
@andrewcking I meant as projects, but visually I see on your sample that there is a difference...
How could it be so slow?
On which device and Android version have you tested it?

**Comment by andrewcking** [comment]
I've tested on nexus 4,5,5x and galaxy s8,s10 and then also on OnePlus 6. It is noticeable on all for my use case. I don't know the android version of each device but it ranges from android 5 on the nexus 4 to android 9 on the galaxy devices. Note these gifs are all local on the device (not loading from the internet).

I don't have a project implementation I can share but my glide calls between 3 and 4 are pretty similar

3.7
```
Glide.with(mActivity).load(images.get(position))
        .crossFade()
        .placeholder(R.drawable.placeholder)
        .into(holder.image);
```

4.+
```
RequestOptions options = new RequestOptions()
         .placeholder(R.drawable.placeholder);
Glide.with(mActivity).load(images.get(position))
        .transition(withCrossFade())
        .apply(options)
        .into(holder.image);
```

**Comment by AndroidDeveloperLB** [comment]
@andrewcking OK I've tested it myself on LG G5 with Android 8.

Attached sample project, using what you guys wrote.

[MyApplication.zip](https://github.com/bumptech/glide/files/3174823/MyApplication.zip)

While I don't see much issue on how fluid it plays, here are the results in CPU usage via the profiler:

4.9.0:

![image](https://user-images.githubusercontent.com/5357526/57655114-88cf7e80-75de-11e9-9533-0d5754944009.png)


3.7.0:

![image](https://user-images.githubusercontent.com/5357526/57656533-48253480-75e1-11e9-8235-872c6d4261e4.png)

How did you perform the test?

**Comment by droidluv** [comment]
Wow was just wondering how good Glide performance was with gif's and this thread just opened my eyes, guess I'll jump onto the Fresco bandwagon for smooth gif playback

**Comment by AndroidDeveloperLB** [comment]
@droidluv Where exactly? Please show a sample showing the issue

**Comment by Zhuinden** [comment]
@droidluv technically, Glide 3.7.2 was ok

**Comment by droidluv** [comment]
> @droidluv Where exactly? Please show a sample showing the issue

I don't have a sample anymore, I encountered a stutter/slowdown when my backend started sending a gif accidentally instead of an image, on a Galaxy On6, and I am using the latest version of Glide (4.9.0), I am hesitant to add Fresco though because of it's much larger size, but saw above that Fresco performed better than Glide with its native code usage, I'll try to generate samples soon enough to compare because a new requirement to support gif playback is incoming.

**Comment by AndroidDeveloperLB** [comment]
Can you please put a sample of the best Glide , the newest one, and of Fresco?

**Comment by droidluv** [comment]
@AndroidDeveloperLB I am attaching the videos below differentiating Fresco and Glide, this slowdown seems specific to only certain devices like I have encountered the Glide slowdown in two Samsung phones but works fine on everything else including other Samsung phones like the Galaxy S7, even though Fresco gives a uniform GIF experience throughout every device, Glide has better performance advantages when dealing with complex images and recyclerviews, like fresco doesn't even support PNG resizing (we can do some downsampling but..) when dealing with complex lists of images and gifs Glide wins hands down, I just switched back to Glide even though I might have to deal with slowed down GIFs in some devices

This comparison is between the newest glide and the latest version of Fresco

[Glide and Fresco comparison.zip](https://github.com/bumptech/glide/files/3612319/Glide.and.Fresco.comparison.zip)

**Comment by AndroidDeveloperLB** [comment]
@droidluv This is of the newest Glide, right?
And you also tried the version of Glide that works fine?
Can you please share the sample code too?

**Comment by droidluv** [comment]
@AndroidDeveloperLB Its on the latest Glide 4.9.0, I haven't tried the 3.x glide because I didn't want to go back to an older version, the code is just a basic vertical Recyclerview with horizontal list's of GIF's and PNG's, Also please do note my point, the slowdown happens only on CERTAIN devices, one such model is the Galaxy J7 Pro, with Android Pie(Android version 9) and OneUI 1.1 and is the stock OS, if you want I can try and generate logs from that device, if I could but I suppose the video should talk for itself, the other device was also a Samsung, a Samsung Galaxy A series device, I can get more info about that device as well if needed, and I'm fairly sure the slowdown is happening on some other devices as well, Fresco seemingly works smoothly everywhere though regarding GIF playback.

**Comment by AndroidDeveloperLB** [comment]
@droidluv Please share the project of Fresco.
I want to investigate it.

**Comment by mertceyhan** [comment]
I found a case. Gif is showing slowly when I using "wrap_content" for ImageView width and height. There is no problem if I use with static width and height such as "80dp" Try this.

```
<androidx.appcompat.widget.AppCompatImageView
        android:id="@+id/imageViewLoadingGif"
        android:layout_width="80dp"
        android:layout_height="80dp" />
```

Also, Gif loader this (Glide 4.10.0);
 Glide
                    .with(this)
                    .asGif()
                    .load(R.raw.loading_animation)
                    .into(this)

**Comment by gmikhail** [comment]
I have similar problem with high frame rate GIFs (50fps). 30fps GIFs play as they should. For comparison - MIUI Gallery plays all GIFs smoothly.

Glide 4.13.2
Poco X3 NFC (Android 11)
GIF details: 50fps, 480x640px, 10 MB, looped

**Comment by cwaliimran** [comment]
Step 1. Make your gif height/width as small as possible, remove extra large background
Step 2. Imageview in XML must have static height/width (rather than 0dp/0dp in constraint layout), it will solve the problem

**Comment by saqib-mirza-confiz** [comment]
None of the above solutions other than the following comment https://github.com/bumptech/glide/issues/2471#issuecomment-1521305640 
Setting fixed height & width of the ImageView worked & resolves the issue of speed

**Comment by rastin-py** [comment]
Setting fixed height & width worked for me. Don't use wrap_content or 0dp in constraint layout.
